modules/free/src/main/scala/doobie/free/Primitive.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2013-2020 Rob Norris and Contributors
+// This software is licensed under the MIT License (MIT).
+// For more information see LICENSE or https://opensource.org/licenses/MIT
+
+package doobie.free
+
+import cats.data.Kleisli
+import cats.effect.Async
+import doobie.WeakAsync
+
+sealed trait Primitive[M[_]] {
+  def primitive[J, A](f: J => A): Kleisli[M, J, A]
+}
+
+object Primitive {
+  case class Default[M[_]]()(implicit asyncM: WeakAsync[M]) extends Primitive[M] {
```

Comment:
In general, it is recommended to avoid using case classes in publicly available libraries unless absolutely necessary. I guess, the primary reason for that here was to get a static `apply` method. Totally understandable it is, but case classes generate a lot of other stuff which makes them difficult to deal with in terms of maintaining binary compatibility. For example: a `copy` method, inheriting `Product`, and the last but not least a notorious `unapply` static method.

A regular class should work just as well here, and if the static `apply` is desirable, it's better to add it manually.

Reply:
The same is applicable to the `case class Cancelable` below.